File: src/main/java/dev/obscuria/obscure_tooltips/client/registry/ResourceKind.java
package dev.obscuria.obscure_tooltips.client.registry;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import dev.obscuria.obscure_tooltips.ObscureTooltips;
import dev.obscuria.obscure_tooltips.client.tooltip.TooltipDefinition;
import dev.obscuria.obscure_tooltips.client.tooltip.TooltipLabel;
import dev.obscuria.obscure_tooltips.client.tooltip.TooltipStyle;
import dev.obscuria.obscure_tooltips.client.tooltip.element.effect.TooltipEffect;
import dev.obscuria.obscure_tooltips.client.tooltip.element.frame.TooltipFrame;
import dev.obscuria.obscure_tooltips.client.tooltip.element.icon.TooltipIcon;
import dev.obscuria.obscure_tooltips.client.tooltip.element.panel.TooltipPanel;
import dev.obscuria.obscure_tooltips.client.tooltip.element.slot.TooltipSlot;
import net.minecraft.util.ResourceLocation;

import java.util.function.Function;

public enum ResourceKind {
    PANEL(new Spec<>("panel", "element/panel", TooltipPanel::fromJson, TooltipRegistries.TOOLTIP_PANEL)),
    FRAME(new Spec<>("frame", "element/frame", TooltipFrame::fromJson, TooltipRegistries.TOOLTIP_FRAME)),
    SLOT(new Spec<>("slot", "element/slot", TooltipSlot::fromJson, TooltipRegistries.TOOLTIP_SLOT)),
    ICON(new Spec<>("icon", "element/icon", TooltipIcon::fromJson, TooltipRegistries.TOOLTIP_ICON)),
    EFFECT(new Spec<>("effect", "element/effect", TooltipEffect::fromJson, TooltipRegistries.TOOLTIP_EFFECT)),
    STYLE(new Spec<>("style", "style", TooltipStyle::fromJson, TooltipRegistries.TOOLTIP_STYLE)),
    DEFINITION(new Spec<>("definition", "definition", TooltipDefinition::fromJson, TooltipRegistries.TOOLTIP_DEFINITION)),
    LABEL(new Spec<>("label", "label", TooltipLabel::fromJson, TooltipRegistries.TOOLTIP_LABEL));

    public final Spec<?> spec;

    ResourceKind(Spec<?> spec) {
        this.spec = spec;
    }

    @Desugar
    public record Spec<T>(
            String name,
            String directory,
            Function<JsonObject, T> parser,
            ResourceRegistry<T> registry) {

        public String resourceDir() {
            return "tooltips/" + directory;
        }

        public void onReloadStart() {
            registry.onReloadStart();
        }

        public void load(ResourceLocation key, JsonObject json) {
            try {
                registry.register(key, parser.apply(json));
            } catch (Exception exception) {
                ObscureTooltips.LOGGER.error("Failed to register {} with key {}: {}", name, key, exception.getMessage());
            }
        }

        public void onReloadEnd() {
            registry.onReloadEnd();
            ObscureTooltips.LOGGER.info("Loaded {} resources from {}", registry.total(), resourceDir());
        }
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/registry/ResourceKind.kt
package dev.obscuria.tooltips.registry

import com.google.gson.JsonElement
import com.mojang.serialization.Codec
import com.mojang.serialization.JsonOps
import dev.obscuria.tooltips.ObscureTooltips.LOGGER
import dev.obscuria.tooltips.client.TooltipDefinition
import dev.obscuria.tooltips.client.TooltipLabel
import dev.obscuria.tooltips.client.TooltipStyle
import dev.obscuria.tooltips.client.element.effect.ITooltipEffect
import dev.obscuria.tooltips.client.element.frame.ITooltipFrame
import dev.obscuria.tooltips.client.element.icon.ITooltipIcon
import dev.obscuria.tooltips.client.element.panel.ITooltipPanel
import dev.obscuria.tooltips.client.element.slot.ITooltipSlot
import net.minecraft.resources.ResourceLocation

enum class ResourceKind(val spec: Spec<*>) {
    PANEL(Spec("panel", "element/panel", ITooltipPanel.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_PANEL)),
    FRAME(Spec("frame", "element/frame", ITooltipFrame.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_FRAME)),
    SLOT(Spec("slot", "element/slot", ITooltipSlot.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_SLOT)),
    ICON(Spec("icon", "element/icon", ITooltipIcon.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_ICON)),
    EFFECT(Spec("effect", "element/effect", ITooltipEffect.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_EFFECT)),
    STYLES(Spec("style", "style", TooltipStyle.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_STYLE)),
    DEFINITIONS(Spec("definition", "definition", TooltipDefinition.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_DEFINITION)),
    LABEL(Spec("label", "label", TooltipLabel.DIRECT_CODEC, TooltipsRegistries.Resource.TOOLTIP_LABEL));

    data class Spec<T>(
        val name: String,
        val directory: String,
        val codec: Codec<T>,
        val registry: ResourceRegistry<T>
    ) {

        fun resourceDir(): String {
            return "tooltips/${directory}"
        }

        fun onReloadStart() {
            registry.onReloadStart()
        }

        fun load(id: ResourceLocation, element: JsonElement) {
            val result = codec.decode(JsonOps.INSTANCE, element)
            result.result().ifPresent { registry.register(id, it.first) }
            result.error().ifPresent { LOGGER.error("Failed to register $name with id $id: ${it.message()}") }
        }

        fun onReloadEnd() {
            registry.onReloadEnd()
            LOGGER.info("Loaded {} resources from {}", registry.total, resourceDir())
        }
    }
}